File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root: return False
        
        def helper(node, leftChild, rightChild ):
            if not node: return True
            
            elif node.val <= leftChild or node.val >= rightChild:
                return False
            
            else: return helper(node.left, leftChild, node.val) and helper(node.right, node.val, rightChild)
            
        
        return helper(root,float('-inf') ,float('inf') )
        

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    /*
    Always left child < root.val < right child
    */
    if(!root) return false;
    if(root.length === 1) return true;
    
    return helper(root, -Infinity, Infinity);
    
    function helper(node, minVal, maxVal){
        if(!node) return true;
        
        if((minVal < node.val) && (node.val < maxVal)){
            return helper(node.left,  minVal, node.val) && helper(node.right, node.val, maxVal)
        } return false
    }
};